### PR TITLE
spork: implement spork caching

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
I was looking through the flow to check if a spork is active or not, and based on my view, it seems extremely wasteful, especially considering that checking if a spork is activate happens extremely often (for example it happens in almost every single IS method).

So this implements caching for spork values and whether a spork is active. 

Whenever a new spork of a given SporkId is processed, all related caching is cleared.

Currently, spork activity is only cached if the activity is "true" this is because false values could be based  on a timestamp, which means we can't really cache false values.

Maybe this change isn't actually really an improvement... LMK

### CODE REVIEW

File: src/spork.cpp [hunk 1]
```diff
@@ -36,12 +36,18 @@ CSporkManager::CSporkManager()
...
     LOCK(cs);
 
     if (!mapSporksActive.count(nSporkID)) return false;
 
+    auto it = mapSporksCachedValues.find(nSporkID);
+    if (it != mapSporksCachedValues.end()) {
+        nActiveValueRet = it->second;
+        return true;
+    }
+
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/spork.cpp:50]: Actually, nvm :)

**Review by UdjinM6 — DISMISSED** [review]
utACK

**Review by xdustinface — CHANGES_REQUESTED** [review]
Two suggestions, see b0f1540e0b561ada933343ff7c79daa354033982 and ee86a0e20ef358aca3fee5593c8fb4a02c73d946 (you dropped the const there)

**Comment by PastaPastaPasta** [comment]
Cherry picked, also, what's the point of making something const, if it's not really const?

Why not make everything mutable, and everything const?

**Comment by xdustinface** [comment]
> what's the point of making something const, if it's not really const?

It its still `const` in general, this only has an impact on members marked `mutable`, the point is IMO that you can still protect other members from getting changed accidentally in a method which shouldn't change anything at all like for example a getter usually just returns a value and should be `const` but you can still allow it to modify specific members if there is a reason like a cache for example by making them `mutable`.

```
class Object {
    int a;
    mutable int b;
    void Magic() const
    {
        ++a; // Not allowed
        ++b; // Allowed
    }
};
```

> Why not make everything mutable, and everything const?

If you do this it's basically like don't using `const`/`mutable` at all, right? So what the point there? 😄

**Review by xdustinface — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**